File: CSharp/Eta/Applications/VAProvider/ProviderCmdLineParser.cs
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license      --
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.  --
 *|                See the project's LICENSE.Md for details.                  --
 *|           Copyright (C) 2022-2023 Refinitiv. All rights reserved.              --
 *|-----------------------------------------------------------------------------
 */

using LSEG.Eta.Example.VACommon;
using LSEG.Eta.Transports;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LSEG.Eta.ValueAdd.Provider
{
    internal class ProviderCmdLineParser : ICommandLineParser
    {
		public string? PortNo { get; set; }
		public string? InterfaceName { get; set; }
		public string? ServiceName { get; set; }
		public int ServiceId { get; set; }
		public int Runtime { get; set; } = 1200; // default runtime is 1200 seconds
		public bool EnableXmlTracing { get; set; }
		public bool EnableRtt { get; set; }
		public string? KeyCertificate { get; set; } = null;
		public string? PrivateKey { get; set; } = null;
		public ConnectionType ConnectionType { get; set; } = ConnectionType.SOCKET;

		public bool ParseArgs(string[] args)
		{
			try
			{
				int argsCount = 0;

				while (argsCount < args.Length)
				{
					if ("-p".Equals(args[argsCount]))
					{
						PortNo = args[++argsCount];
						++argsCount;
					}
					else if ("-i".Equals(args[argsCount]))
					{
						InterfaceName = args[++argsCount];
						++argsCount;
					}
					else if ("-s".Equals(args[argsCount]))
					{
						ServiceName = args[++argsCount];
						++argsCount;
					}
					else if ("-id".Equals(args[argsCount]))
					{
						ServiceId = int.Parse(args[++argsCount]);
						++argsCount;
					}
					else if ("-runtime".Equals(args[argsCount]))
					{
						Runtime = int.Parse(args[++argsCount]);
						++argsCount;
					}
					else if ("-x".Equals(args[argsCount]))
					{
						EnableXmlTracing = true;
						++argsCount;
					}
					else if ("-rtt".Equals(args[argsCount]))
					{
						EnableRtt = true;
						++argsCount;
					}
					else if ("-c".Equals(args[argsCount]))
					{
						++argsCount;
						if ("socket".Equals(args[argsCount]))
						{
							ConnectionType = ConnectionType.SOCKET;
						}
						else if ("encrypted".Equals(args[argsCount]))
						{
							ConnectionType = ConnectionType.ENCRYPTED;
						}
						else
						{
							Console.WriteLine("\nUnrecognized connection type...\n");
							return false;
						}
						++argsCount;
					}
					else if ("-cert".Equals(args[argsCount]))
					{
						KeyCertificate = args[++argsCount];
						++argsCount;
					}
					else if ("-keyfile".Equals(args[argsCount]))
					{
						PrivateKey = args[++argsCount];
						++argsCount;
					}
					else // unrecognized command line argument
					{
						Console.WriteLine("\nUnrecognized command line argument...\n");
						return false;
					}
				}
			}
			catch (Exception e)
			{
				Console.WriteLine($"\nInvalid command line arguments: {e.Message}");
				return false;
			}

			return true;
		}

		public void PrintUsage()
		{
			Console.WriteLine("Usage: Provider or\nProvider [-p <port number>] [-i <interface name>] [-s <service name>] [-id <service ID>] [-runtime <seconds>]" +
							   "\n -p server port number (defaults to 14002)\n" +
							   "\n -i interface name (defaults to null)\n" +
							   "\n -s service name (defaults to DIRECT_FEED)\n" +
							   "\n -id service id (defaults to 1)\n" +
							   "\n -x provides an XML trace of messages\n" +
							   "\n -runtime application runtime in seconds (default is 1200)" +
							   "\n -rtt application (provider) supports calculation of Round Trip Latency" +
							   "\n -c Provider connection type.  Either \"socket\" or \"encrypted\"" +
							   "\n -key the server private key file" +
							   "\n -cert the server certificate file");
		}


	}
}

File: Java/Eta/TestTools/QATools/vaprovider-TsPost-001/ProviderCmdLineParser.java
/*|-----------------------------------------------------------------------------
 *|            This source code is provided under the Apache 2.0 license
 *|  and is provided AS IS with no warranty or guarantee of fit for purpose.
 *|                See the project's LICENSE.md for details.
 *|           Copyright (C) 2022,2024 LSEG. All rights reserved.
 *|-----------------------------------------------------------------------------
 */

package com.refinitiv.eta.valueadd.examples.provider;

import com.refinitiv.eta.transport.ConnectionTypes;
import com.refinitiv.eta.valueadd.examples.common.CommandLineParser;

class ProviderCmdLineParser implements CommandLineParser
{
	private String portNo;
	private String interfaceName;
	private String serviceName;
	private int serviceId;
	private int runtime = 1200; // default runtime is 1200 seconds
	private boolean enableXmlTracing;
	private boolean cacheOption;
	private boolean enableRtt;
	private String keyfile = null;
	private String keypasswd = null;
	private int connType = ConnectionTypes.SOCKET;
	private String protocolList = "rssl.rwf, tr_json2, rssl.json.v2";
	private boolean sendNack;
	private boolean rejectLogin;
	private boolean rejectTsLogin;

	@Override
	public boolean parseArgs(String[] args)
	{
		try
		{
			int argsCount = 0;

			while (argsCount < args.length)
			{
				if ("-p".equals(args[argsCount]))
				{
					portNo = args[++argsCount];
					++argsCount;
				}
				else if ("-i".equals(args[argsCount]))
				{
					interfaceName = args[++argsCount];
					++argsCount;
				}
				else if ("-s".equals(args[argsCount]))
				{
					serviceName = args[++argsCount];
					++argsCount;
				}
				else if ("-id".equals(args[argsCount]))
				{
					serviceId = Integer.parseInt(args[++argsCount]);
					++argsCount;
				}
				else if ("-runtime".equals(args[argsCount]))
				{
					runtime = Integer.parseInt(args[++argsCount]);
					++argsCount;
				}
				else if ("-x".equals(args[argsCount]))
				{
					enableXmlTracing =  true;
					++argsCount;
				}
				else if ("-cache".equals(args[argsCount]))
				{
					cacheOption =  true;
					++argsCount;
				} else if ("-rtt".equals(args[argsCount])) {
					enableRtt = true;
					++argsCount;
				} else if ("-c".equals(args[argsCount])) {
					++argsCount;
					if("socket".equals(args[argsCount]))
					{
						connType = ConnectionTypes.SOCKET;
					}
					else if("encrypted".equals(args[argsCount]))
					{
						connType = ConnectionTypes.ENCRYPTED;
					}
					else
					{
						System.out.println("\nUnrecognized connection type...\n");
						return false;
					}
					++argsCount;
				} else if ("-keyfile".equals(args[argsCount])) {
					keyfile = args[++argsCount];
					++argsCount;
				} else if ("-keypasswd".equals(args[argsCount])) {
					keypasswd = args[++argsCount];
					++argsCount;
				} else if ("-pl".equals(args[argsCount])) {
					protocolList = args[++argsCount];
					++argsCount;
				}
				else if ("-sendNack".equals(args[argsCount]))
				{
					sendNack = true;
					++argsCount;
				}
				else if ("-rejectLogin".equals(args[argsCount]))
				{
					rejectLogin = true;
					++argsCount;
				}
				else if ("-rejectTsLogin".equals(args[argsCount]))
				{
					rejectTsLogin = true;
					++argsCount;
				}
				else // unrecognized command line argument
				{
					System.out.println("\nUnrecognized command line argument...\n");
					return false;
				}
			}
		}
		catch (Exception e)
		{
			System.out.println("\nInvalid command line arguments...");
			return false;
		}

		return true;
	}

	@Override
	public void printUsage()
	{
		System.out.println("Usage: Provider or\nProvider [-p <port number>] [-i <interface name>] [-s <service name>] [-id <service ID>] [-runtime <seconds>]" +
						   "\n -p server port number (defaults to 14002)\n" +
						   "\n -i interface name (defaults to null)\n" +
						   "\n -s service name (defaults to DIRECT_FEED)\n" +
						   "\n -id service id (defaults to 1)\n" +
						   "\n -x provides an XML trace of messages\n" +
						   "\n -cache turn on the cache feature of the application\n" +
						   "\n -pl protocol list (defaults to rssl.rwf, tr_json2, rssl.json.v2)\n" +
						   "\n -runtime application runtime in seconds (default is 1200)" +
						   "\n -rtt application (provider) supports calculation of Round Trip Latency" +
						   "\n -c Provider connection type.  Either \"socket\" or \"encrypted\"" +
						   "\n -keyfile jks encoded keyfile for Encrypted connections" +
						   "\n -keypasswd password for keyfile" +
				           "\n -sendNack send negative acks if post message is received (defaults to false)" +
				           "\n -rejectLogin to reject main login request (defaults to false)" +
				           "\n -rejectTsLogin to reject the tunnel stream login request (defaults to false)");
	}

	String portNo()
	{
		return portNo;
	}

	String interfaceName()
	{
		return interfaceName;
	}

	String serviceName()
	{
		return serviceName;
	}

	int serviceId()
	{
		return serviceId;
	}

	int runtime()
	{
		return runtime;
	}

	boolean enableXmlTracing()
	{
		return enableXmlTracing;
	}

	boolean cacheOption()
	{
		return cacheOption;
	}

	boolean enableRtt()
	{
		return enableRtt;
	}

	int connectionType()
	{
		return connType;
	}

	String keyfile()
	{
		return keyfile;
	}

	String keypasswd()
	{
		return keypasswd;
	}

	String protocolList()
	{
		return protocolList;
	}

	boolean sendNack()
	{
		return sendNack;
	}

	boolean rejectLogin() {
		return rejectLogin;
	}

	boolean rejectTsLogin() {
		return rejectTsLogin;
	}
}
